File: dns_lookup.cs
// dns_lookup.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

class DNSResult
{
    public string Domain { get; set; }
    public string Timestamp { get; set; }
    public List<string> AAAA { get; set; } = new List<string>();
    public List<string> PTR { get; set; } = new List<string>();
}

class DNSLookupIPv6
{
    private int timeout = 5000;
    private int threads = 20;
    private List<DNSResult> results = new List<DNSResult>();
    private object locker = new object();

    public async Task<List<DNSResult>> LookupBatchAsync(List<string> domains)
    {
        Console.WriteLine($"🔍 Выполняем lookup для {domains.Count} доменов...");
        var start = DateTime.Now;

        var tasks = new List<Task>();
        var semaphore = new SemaphoreSlim(threads);

        foreach (var domain in domains.Where(d => !string.IsNullOrWhiteSpace(d)))
        {
            await semaphore.WaitAsync();
            tasks.Add(Task.Run(async () =>
            {
                try
                {
                    var result = await LookupDomainAsync(domain.Trim());
                    if (result != null)
                    {
                        lock (locker)
                        {
                            results.Add(result);
                        }
                        Console.WriteLine($"✅ {domain.Trim()} — выполнено");
                    }
                }
                finally
                {
                    semaphore.Release();
                }
            }));
        }

        await Task.WhenAll(tasks);
        Console.WriteLine($"⏱️ Время выполнения: {(DateTime.Now - start).TotalSeconds:F2} сек.");
        return results;
    }

    private async Task<DNSResult> LookupDomainAsync(string domain)
    {
        var result = new DNSResult
        {
            Domain = domain,
            Timestamp = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss.fffZ")
        };

        try
        {
            // Проверяем, является ли строка IPv6-адресом
            if (IPAddress.TryParse(domain, out var ip) && ip.AddressFamily == System.Net.Sockets.AddressFamily.InterNetworkV6)
            {
                var ptr = await GetPTRAsync(domain);
                result.PTR = ptr.Any() ? ptr : new List<string> { "не найден" };
                return result;
            }

            var entries = await Dns.GetHostEntryAsync(domain);
            var ipv6Addresses = entries.AddressList.Where(a => a.AddressFamily == System.Net.Sockets.AddressFamily.InterNetworkV6).ToList();

            result.AAAA = ipv6Addresses.Select(a => a.ToString()).ToList();

            if (ipv6Addresses.Any())
            {
                foreach (var addr in ipv6Addresses)
                {
                    var ptr = await GetPTRAsync(addr.ToString());
                    result.PTR.AddRange(ptr);
                }
            }
            if (!result.PTR.Any())
            {
                result.PTR.Add("не найден");
            }
        }
        catch
        {
            result.PTR.Add("не найден");
        }

        return result;
    }

    private async Task<List<string>> GetPTRAsync(string ip)
    {
        try
        {
            var entry = await Dns.GetHostEntryAsync(ip);
            return new List<string> { entry.HostName };
        }
        catch
        {
            return new List<string>();
        }
    }

    public void PrintTable()
    {
        if (results.Count == 0)
        {
            Console.WriteLine("Нет результатов.");
            return;
        }

        Console.WriteLine("\n" + new string('=', 90));
        Console.WriteLine($"{"Домен/IP",-25} {"AAAA",-35} {"PTR",-25}");
        Console.WriteLine(new string('-', 90));
        foreach (var r in results)
        {
            var aaaa = string.Join("; ", r.AAAA);
            if (string.IsNullOrEmpty(aaaa)) aaaa = "—";
            if (aaaa.Length > 35) aaaa = aaaa.Substring(0, 35) + "...";
            var ptr = string.Join("; ", r.PTR);
            if (ptr.Length > 25) ptr = ptr.Substring(0, 25) + "...";
            Console.WriteLine($"{r.Domain,-25} {aaaa,-35} {ptr,-25}");
        }
        Console.WriteLine(new string('=', 90));
    }

    public void SaveJSON(string filename)
    {
        var options = new JsonSerializerOptions { WriteIndented = true };
        var json = JsonSerializer.Serialize(results, options);
        File.WriteAllText(filename, json);
        Console.WriteLine($"💾 Сохранено JSON: {filename}");
    }

    public void SaveCSV(string filename)
    {
        if (results.Count == 0) return;
        var lines = new List<string>();
        lines.Add("Domain,Timestamp,AAAA,PTR");
        foreach (var r in results)
        {
            lines.Add($"{r.Domain},{r.Timestamp},{string.Join("; ", r.AAAA)},{string.Join("; ", r.PTR)}");
        }
        File.WriteAllLines(filename, lines);
        Console.WriteLine($"💾 Сохранено CSV: {filename}");
    }
}

class Program
{
    static async Task Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: dotnet run <domains.txt>");
            Console.WriteLine("   или: dotnet run domain1.com domain2.com");
            return;
        }

        List<string> domains = new List<string>();

        if (args.Length == 1 && File.Exists(args[0]))
        {
            domains = File.ReadAllLines(args[0])
                          .Where(line => !string.IsNullOrWhiteSpace(line))
                          .Select(line => line.Trim())
                          .ToList();
        }
        else
        {
            domains = args.ToList();
        }

        if (domains.Count == 0)
        {
            Console.WriteLine("❌ Нет доменов для проверки.");
            return;
        }

        Console.WriteLine("🌐 DNS Lookup IPv6 Pro (C#)");
        Console.WriteLine($"📂 Загружено {domains.Count} доменов.");

        var lookup = new DNSLookupIPv6();
        var results = await lookup.LookupBatchAsync(domains);

        if (results.Count > 0)
        {
            lookup.PrintTable();
            lookup.SaveJSON("dns_results.json");
            lookup.SaveCSV("dns_results.csv");
        }
        else
        {
            Console.WriteLine("❌ Не удалось выполнить lookup.");
        }
    }
}

File: dns_lookup.rs
// dns_lookup.rs — Rust версия

use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader, Write};
use std::time::Instant;
use std::sync::Arc;
use tokio::sync::Semaphore;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
struct DNSResult {
    domain: String,
    timestamp: String,
    AAAA: Vec<String>,
    PTR: Vec<String>,
}

impl DNSResult {
    fn new(domain: String) -> Self {
        DNSResult {
            domain,
            timestamp: chrono::Utc::now().to_rfc3339(),
            AAAA: Vec::new(),
            PTR: Vec::new(),
        }
    }
}

fn is_ipv6(addr: &str) -> bool {
    addr.parse::<std::net::Ipv6Addr>().is_ok()
}

async fn lookup_domain(domain: &str) -> Option<DNSResult> {
    let domain = domain.trim();
    if domain.is_empty() {
        return None;
    }

    let mut result = DNSResult::new(domain.to_string());

    // Проверяем, является ли строка IPv6-адресом
    if is_ipv6(domain) {
        if let Ok(ips) = tokio::net::lookup_host((domain, 0)).await {
            for ip in ips {
                if ip.is_ipv6() {
                    if let Ok(ptr) = tokio::net::lookup_host((ip.ip().to_string(), 0)).await {
                        for p in ptr {
                            result.PTR.push(p.ip().to_string());
                        }
                    }
                }
            }
        }
        if result.PTR.is_empty() {
            result.PTR.push("не найден".to_string());
        }
        return Some(result);
    }

    // AAAA lookup
    if let Ok(ips) = tokio::net::lookup_host((domain, 0)).await {
        for ip in ips {
            if ip.is_ipv6() {
                result.AAAA.push(ip.ip().to_string());
            }
        }
    }

    // PTR для каждого AAAA адреса
    if !result.AAAA.is_empty() {
        for ip in &result.AAAA {
            if let Ok(ips) = tokio::net::lookup_host((ip.as_str(), 0)).await {
                for p in ips {
                    result.PTR.push(p.ip().to_string());
                }
            }
        }
    }
    if result.PTR.is_empty() {
        result.PTR.push("не найден".to_string());
    }

    Some(result)
}

async fn lookup_batch(domains: Vec<String>, threads: usize) -> Vec<DNSResult> {
    println!("🔍 Выполняем lookup для {} доменов...", domains.len());
    let start = Instant::now();

    let semaphore = Arc::new(Semaphore::new(threads));
    let mut handles = Vec::new();

    for domain in domains {
        if domain.trim().is_empty() {
            continue;
        }
        let sem = semaphore.clone();
        let domain = domain.trim().to_string();
        let handle = tokio::spawn(async move {
            let _permit = sem.acquire().await.unwrap();
            let result = lookup_domain(&domain).await;
            if let Some(r) = result {
                println!("✅ {} — выполнено", domain);
                r
            } else {
                println!("⚠️ {} — ошибка", domain);
                DNSResult::new(domain)
            }
        });
        handles.push(handle);
    }

    let mut results = Vec::new();
    for handle in handles {
        if let Ok(result) = handle.await {
            if !result.AAAA.is_empty() || result.PTR.iter().any(|p| p != "не найден") {
                results.push(result);
            }
        }
    }

    println!("⏱️ Время выполнения: {:.2} сек.", start.elapsed().as_secs_f64());
    results
}

fn print_table(results: &[DNSResult]) {
    if results.is_empty() {
        println!("Нет результатов.");
        return;
    }

    println!("\n{}", "=".repeat(90));
    println!("{:<25} {:<35} {:<25}", "Домен/IP", "AAAA", "PTR");
    println!("{}", "-".repeat(90));
    for r in results {
        let aaaa = r.AAAA.join("; ");
        let aaaa = if aaaa.is_empty() { "—".to_string() } else { aaaa };
        let aaaa = if aaaa.len() > 35 { &aaaa[..35] } else { &aaaa };
        let ptr = r.PTR.join("; ");
        let ptr = if ptr.len() > 25 { &ptr[..25] } else { &ptr };
        println!("{:<25} {:<35} {:<25}", r.domain, aaaa, ptr);
    }
    println!("{}", "=".repeat(90));
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        println!("Usage: cargo run -- <domains.txt>");
        println!("   или: cargo run -- domain1.com domain2.com");
        return Ok(());
    }

    let mut domains = Vec::new();

    if args.len() == 2 && std::path::Path::new(&args[1]).exists() {
        let file = File::open(&args[1])?;
        let reader = BufReader::new(file);
        for line in reader.lines() {
            let line = line?;
            let domain = line.trim();
            if !domain.is_empty() {
                domains.push(domain.to_string());
            }
        }
    } else {
        for arg in args.iter().skip(1) {
            if !arg.is_empty() {
                domains.push(arg.clone());
            }
        }
    }

    if domains.is_empty() {
        println!("❌ Нет доменов для проверки.");
        return Ok(());
    }

    println!("🌐 DNS Lookup IPv6 Pro (Rust)");
    println!("📂 Загружено {} доменов.", domains.len());

    let results = lookup_batch(domains, 20).await;

    if !results.is_empty() {
        print_table(&results);

        let json = serde_json::to_string_pretty(&results)?;
        let mut file = File::create("dns_results.json")?;
        file.write_all(json.as_bytes())?;
        println!("💾 Сохранено JSON: dns_results.json");

        let mut csv_file = File::create("dns_results.csv")?;
        writeln!(csv_file, "Domain,Timestamp,AAAA,PTR")?;
        for r in &results {
            writeln!(csv_file, "{},{},{},{}",
                r.domain, r.timestamp,
                r.AAAA.join("; "),
                r.PTR.join("; ")
            )?;
        }
        println!("💾 Сохранено CSV: dns_results.csv");
    } else {
        println!("❌ Не удалось выполнить lookup.");
    }

    Ok(())
}
